File: src/main/java/org/emes/RecurringParser.java
package org.emes;

import java.util.List;
import java.util.function.Function;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

//create go shopping every thu //MODE EVERY WEEK
//create go shopping every 1st //MODE EVERY MONTH DAY
//create go shopping every 2 months //MODE EVERY X MONTHS - NEXT STARTING
//create go shopping every 3 weeks //MODE EVERY X WEEKS - NEXT STARTING
//create go shopping every 10 days // MODE EVERY X DAYS - NEXT STARTING
public class RecurringParser {

  private static final Logger LOG = LogManager.getLogger();

  private final static Pattern EVERY_DAY_OF_WEEK = Pattern.compile(
      "\\bevery (mon|tue|wed|thu|fri|sat|sun)\\b");
  private final static Pattern EVERY_DAY_OF_MONTH = Pattern.compile(
      "\\bevery (\\d)+ \\b"); //TODO
  private final static Pattern EVERY_X_DAYS = Pattern.compile("\\bevery (\\d) day(s)?\\b");
  private final static Pattern EVERY_X_WEEKS = Pattern.compile("\\bevery (\\d) week(s)?\\b");
  private final static Pattern EVERY_X_MONTHS = Pattern.compile("\\bevery (\\d) month(s)?\\b");

  private final List<Pair<Pattern, Function<MatchResult, Recurring>>> patterns = List.of(
      new Pair<>(EVERY_DAY_OF_WEEK, this::handleEveryDayOfWeek),
      new Pair<>(EVERY_DAY_OF_MONTH, this::handleEveryDayOfMonth),
      new Pair<>(EVERY_X_DAYS, this::handleEveryXDays),
      new Pair<>(EVERY_X_WEEKS, this::handleEveryXWeeks),
      new Pair<>(EVERY_X_MONTHS, this::handleEveryXMonths)
  );

  public Recurring parse(String command) {
    return patterns.stream()
        .filter(it -> it.a().matcher(readLine).find())
        .findFirst()
        .map(it -> {
          LOG.info("Use parser: {}", it.a());
          var matchResult = it.a().matcher(readLine);
          matchResult.find();
          return it.b().apply(matchResult.toMatchResult());
        })
        .orElse(null);
  }

  private Recurring handleEveryDayOfWeek(MatchResult matchResult) {
    return new Recurring(RecurringMode.NONE, null, null, null, null);
  }

  private Recurring handleEveryDayOfMonth(MatchResult matchResult) {
    return new Recurring(RecurringMode.NONE, null, null, null, null);
  }

  private Recurring handleEveryXDays(MatchResult matchResult) {
    return new Recurring(RecurringMode.NONE, null, null, null, null);
  }

  private Recurring handleEveryXWeeks(MatchResult matchResult) {
    return new Recurring(RecurringMode.NONE, null, null, null, null);
  }

  private Recurring handleEveryXMonths(MatchResult matchResult) {
    return new Recurring(RecurringMode.NONE, null, null, null, null);
  }
}

File: src/main/java/org/emes/parser/RecurringParser.kt
package org.emes.parser

import io.github.oshai.kotlinlogging.KotlinLogging
import org.emes.parser.ParseResult
import org.emes.Recurring
import org.emes.RecurringMode

class RecurringParser {

    private val logger = KotlinLogging.logger {}

    private val everyDayOfWeek = Regex("\\bin (\\d)+ (day|week|month)s?\\b")
    private val everyDayOfMonth = Regex("\\bin (\\d)+ (day|week|month)s?\\b")
    private val everyXDays = Regex("\\bin (\\d)+ (day|week|month)s?\\b")
    private val everyXWeeks = Regex("\\bin (\\d)+ (day|week|month)s?\\b")
    private val everyXMonths = Regex("\\bin (\\d)+ (day|week|month)s?\\b")

    private val patternWithHandlers = listOf<RegexWithHandler>(
        RegexWithHandler(everyDayOfWeek) { result -> handleEveryDayOfWeek(result) },
        RegexWithHandler(everyDayOfMonth) { result -> handleEveryDayOfMonth(result) },
        RegexWithHandler(everyXDays) { result -> handleEveryXDays(result) },
        RegexWithHandler(everyXWeeks) { result -> handleEveryXWeeks(result) },
        RegexWithHandler(everyXMonths) { result -> handleEveryXMonths(result) }
    )

    fun parse(command: String): ParseResult<Recurring>? {
        return patternWithHandlers
            .mapNotNull { patternWithHandler ->
                patternWithHandler.regex.findAll(command).lastOrNull()?.let { match ->
                    var recurring = patternWithHandler.handler(match)
                    logger.info { "Use parser: ${patternWithHandler.regex.pattern}" }
                    return ParseResult(recurring, match.range)
                }
            }
            .firstOrNull()
        //TODO handle case when multiple matches
    }

    private fun handleEveryDayOfWeek(matchResult: MatchResult): Recurring {
        return Recurring(RecurringMode.EVERY_DAY_OF_WEEK, null, null, null, null)
    }

    private fun handleEveryDayOfMonth(matchResult: MatchResult): Recurring {
        return Recurring(RecurringMode.EVERY_DAY_OF_WEEK, null, null, null, null)
    }

    private fun handleEveryXDays(matchResult: MatchResult): Recurring {
        return Recurring(RecurringMode.EVERY_DAY_OF_WEEK, null, null, null, null)
    }

    private fun handleEveryXWeeks(matchResult: MatchResult): Recurring {
        return Recurring(RecurringMode.EVERY_DAY_OF_WEEK, null, null, null, null)
    }

    private fun handleEveryXMonths(matchResult: MatchResult): Recurring {
        return Recurring(RecurringMode.EVERY_DAY_OF_WEEK, null, null, null, null)
    }

    private data class RegexWithHandler(
        val regex: Regex,
        val handler: (MatchResult) -> Recurring
    )
}